Infinite loop detection for const evaluation

ecstatic-morse opened the pull request:
Resolves #50637.

An `EvalContext` stores the transient state (stack, heap, etc.) of the MIRI virtual machine while it executing code. As long as MIRI only executes pure functions, we can detect if a program is in a state where it will never terminate by periodically taking a "snapshot" of this transient state and comparing it to previous ones. If any two states are exactly equal, the machine must be in an infinite loop.

Instead of fully cloning a snapshot every time the detector is run, we store a snapshot's hash. Only when a hash collision occurs do we fully clone the interpreter state. Future snapshots which cause a collision will be compared against this clone, causing the interpreter to abort if they are equal.

At the moment, snapshots are not taken until MIRI has progressed a certain amount. After this threshold, snapshots are taken every `DETECTOR_SNAPSHOT_PERIOD` steps. This means that an infinite loop with period `P` will be detected after a maximum of `2 * P * DETECTOR_SNAPSHOT_PERIOD` interpreter steps. The factor of 2 arises because we only clone a snapshot after it causes a hash collision.

rust-highfive commented:
r? @michaelwoerister

(rust_highfive has picked a reviewer for you, use r? to override)

ecstatic-morse commented:
r? @oli-obk

oli-obk commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -33,11 +33,8 @@ pub struct EvalContext<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
     /// Bounds in scope for polymorphic evaluations.
     pub param_env: ty::ParamEnv<'tcx>,
 
-    /// The virtual memory system.
```
The `machine` field also needs to go into the `EvalState`. It's unused in the compiler, but `miri` the tool needs it

ecstatic-morse replied:
Got it, when I first started on this, I assumed that I'd need to provide some custom implementations of common methods on `EvalState`, as well as pass it by reference to something like `HashSet::get`. However, since I implemented custom Eq/Hash on `Frame` and `Memory`, `observe` could simply accept several references to individulal fields.

This means I could revert my first commit and not break everything.

ecstatic-morse replied:
Also, I could use #[derive] more as well as remove a lot of where bounds if I add `Eq + Hash + Clone` to the supertraits of `Machine`. What's rustc's general philosophy on this? I try to avoid needlessly adding supertraits, but here it would save a lot of boilerplate.

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -992,12 +992,12 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> HasMemory<'a, 'mir, 'tcx, M> for Me
     #[inline]
     fn memory_mut(&mut self) -> &mut Memory<'a, 'mir, 'tcx, M> {
-        &mut self.memory
+        self.memory_mut()
```
This feels very infinite recursion dangerous. Or at least surprising to read. Can we just get rid of the `memory_mut` and `memory` inherent methods in favour of this trait impl?

ecstatic-morse replied:
`HasMemory` is private to this crate, and `tools/miri` would need to call these methods.

The solution above would obviate this problem.

oli-obk commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -95,6 +134,52 @@ pub struct Frame<'mir, 'tcx: 'mir> {
+            stmt,
+        } = self;
+
+        ptr::eq(mir, &other.mir)
```
You need to check the `instance` for equality, because it contains the `substs`, which make up the generic arguments of the current frame's function. It's probably not problematic due to the upper frame's `stmt` field, but we should still have this `PartialEq` impl be correct.

This means you also don't have to check the Mir, because the Mir depends directly on the `DefId` inside the `Instance`.

ecstatic-morse replied:
Got it, I've read the rustc book, but some of the finer points eluded me.

oli-obk commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -95,6 +134,52 @@ pub struct Frame<'mir, 'tcx: 'mir> {
+        } = self;
+
+        ptr::eq(mir, &other.mir)
+            && *return_to_block == other.return_to_block // TODO: Are these two necessary?
```
Probably not in a global way (due to `stmt` on the upper stack frame), but still nice to have this impl be very correct.

oli-obk commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -95,6 +134,52 @@ pub struct Frame<'mir, 'tcx: 'mir> {
+            stmt,
+        } = self;
+
+        (mir as *const _ as usize).hash(state);
```
same here, hash the instance, not the mir.

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
+            alloc_map: self.alloc_map.clone(),
+            uninitialized_statics: self.uninitialized_statics.clone(),
+            cur_frame: self.cur_frame.clone(),
+            tcx: self.tcx.clone(),
```
no need for `.clone()` here.

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
+            alloc_kind: self.alloc_kind.clone(),
+            alloc_map: self.alloc_map.clone(),
+            uninitialized_statics: self.uninitialized_statics.clone(),
+            cur_frame: self.cur_frame.clone(),
```
no need for `.clone()` here.

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
+            && *alloc_map == other.alloc_map
+            && *uninitialized_statics == other.uninitialized_statics
+            && *cur_frame == other.cur_frame
+            && ptr::eq(tcx, &other.tcx)
```
no need to compare the `tcx`, there is only ever one, and the `Span` changing is not something we care about.

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
+
+        data.hash(state);
+        cur_frame.hash(state);
+        (tcx as *const _ as usize).hash(state);
```
also no need to hash the tcx.

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
+        cur_frame.hash(state);
+        (tcx as *const _ as usize).hash(state);
+
+        // We ignore some fields which don't change between evaluation steps.
```
You aren't really ignoring them, hashing the `allocations` iterator should get you most of the info.

ecstatic-morse replied:
This ignores `uninitialized_statics` at the moment, which actually can change during execution, but I think will not affect infinite loop detection, since modifying an uninitialized static would cause it to no longer be initialized. I could be way off here though.

oli-obk replied:
That's dangerous. Equality and hashing need to be in sync.  I don't remember the exact rules, but you can read up on that in the hash docs

ecstatic-morse replied:
`uninitialized_statics` (along with another field) is ignored in the hash, but not in the equality check. This maintains the invariant required by `Hash`, that two objects with different hashes never compare as equal.

This should be fixed anyways, but I need to know how to compare `FxHashMap<AllocId, Allocation>`s against one another.

oli-obk replied:
Right. That makes sense

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
+        self.allocations()
+            .map(|allocs| {
+                let mut h = FxHasher::default();
+                allocs.hash(&mut h);
```
So this is problematic. Hashing an `AllocId` (inside a relocation of the alloc) will hash the actual `Id`. This means that if we ever call a function in a loop, since we keep getting new `AllocId`s for the function's locals, we'll never realize we're in a loop (ok we will outside the function, but imagine the function returning a raw pointer to its local, and we keep updating a field in the outer function).

I think it's fine for now, and will probably go away once we get around to implementing local `AllocId`s.

ecstatic-morse replied:
So right now, this hashes *both* the `AllocId` and the `Allocation` (which can be crazy slow btw). Would simply hashing `Allocation` be correct?

I could maybe improve performance by memoizing the hash of an `Allocation`. This would involve setting a `Cell<Option<u64>>` to `None` whenever an `Allocation` is mutated, even when the loop detector isn't running. This might not be too bad, since there's no immediate data dependency on the cached value, and it will often be `None` during regular execution so the branch predictor will be on our side.

oli-obk replied:
We could benchmark it. For now let's do the slow version. We are already in "this might take a while territory"

oli-obk replied:
We might not even have to hash the memory, if we transitively hash through AllocIds by hashing their allocation instead of the id

ecstatic-morse replied:
Agreed, I'm a sucker for a premature optimization :).

> transitively hash through AllocIds by hashing their allocation instead of the id

Can you elaborate on this? I don't quite understand.

oli-obk replied:
So. The frame locals can contain AllocIds. Instead of just hashing these normally, you'd have to hand roll recursive hashing code that, whenever it sees an AllocId, instead of hashing that u64, it loads the allocation from the memory and the hashes that the same way: any encountered AllocIds are instead used to hash their contents.

In order for that not to explode you need to keep around a `HashMap<AllocId, u64>` where the u64 are already hashed Allocation hashes, so you can just hash that hash instead of the entire allocation

ecstatic-morse replied:
Okay, so the recursive part is the (`Relocations`)[https://github.com/rust-lang/rust/blob/441290291fc71dbfa16a372636a59da75649eece/src/librustc/mir/interpret/mod.rs#L480] field of `Allocation`.

ecstatic-morse replied:
I'm caught up now; I'll repeat what you said for my own clarity. 

`alloc_map`, contains all `Allocation`s which are visible to a given `EvalContext` indexed by `AllocId`. The naive approach, comparing the `alloc_map`s with `eq`, will correctly identify some infinite loops. However,  if an object is reallocated, and all pointers to it are updated to point to the new value, the program states are equivalent but our equality function will fail.

Instead, we can look at the program heap as a graph of allocations whose roots are `Value::ByRef`s on the stack and whose edges are the `AllocId`s in `Relocations`. Equality over `EvalSnaphost` `Frame`s is the same except when it comes to these `ByRef` variants. We would push them onto a separate list for each frame, then do a simultaneous BFS/DFS, comparing each `Allocation` along the way. This would make equality invariant over changing `AllocId`s.

It is much easier to implement `Hash` than `PartialEq` with this approach: we simply omit any `AllocId`s--either on the stack in a `Value::ByRef` or the heap in a `Relocations`--from the hash.

I think I will implement this naively at first (not handling object reallocation), then try to come up with a test case where the more robust case would be necessary to detect infinite loops.

ecstatic-morse replied:
Also, I think the result of casting a reference to `usize` during CTFE is still up in the air. If the `AllocId` is used in any way when computing that, the more complex strategy will no longer be correct.

ecstatic-morse replied:
The following test case fails as you predicted.

```rust
#![feature(const_fn, const_let)]

const fn churn() -> *const u8 {
    let x = 0u8;
    &x
}

fn main() {
    let _ = [(); {
        let mut y = 0 as *const u8;
        loop {
            y = churn();
        }
        y as usize
    }];
}
```

There must be a subtler way of triggering this, but I can't come up with one at the moment. I don't think this class of programs is worth implementing a custom equality function which ignores `AllocId`s. Let me know if you disagree.

oli-obk replied:
We can at least check for each AllocId that we hash, whether it is dangling, and then just hash some default value (e.g. `u64::max_value`), unfortunately that requires the same changes as walking down AllocIds...

If you open an issue about the not working example and the explanation of the full fix, we can merge this PR as it is.

ecstatic-morse replied:
I'll open an issue. I plan to keep working on this btw.

bjorn3 commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
     pub tcx: TyCtxtAt<'a, 'tcx, 'tcx>,
 }
 
+impl<'a, 'mir, 'tcx, M> Clone for Memory<'a, 'mir, 'tcx, M>
```
Can this be derived?

oli-obk replied:
Should be possible after the changes I suggested where the `machine` field of the `EvalContext` becomes part of the state.

ecstatic-morse replied:
@bjorn3 Because I chose to refactor the fields of `EvalContext` into a separate struct so that it could have a custom hash impl and be looked up by reference in a hash table. See [my comment](https://github.com/rust-lang/rust/pull/51702#discussion_r197505316) for why this might be overkill. Once you do that, everything that was referring to `ctxt.stack` had to be changed to either `ctxt.stack{,_mut}()` or `ctxt.state.stack`.

ecstatic-morse replied:
I did this because explicitly because deriving `Clone` for `Memory` would add a `where M: Clone` bound to the implementation, even though only `where M::MemoryData: Clone` is needed.

ecstatic-morse commented:
I reverted the `EvalContext` layout changes so that I don't cause so much churn by changing field accesses to method calls everywhere. I can always reapply that patch later on. Now `observe` just takes a reference to each field which is significant for execution state.

[This comment](https://github.com/rust-lang/rust/pull/51702#discussion_r197557032) still needs to be resolved regarding what parts of `Memory` should be hashed.

My next steps are:
* Adding an `EvalErrorKind::InfiniteLoop` variant
* Execute [this](https://github.com/rust-lang/rust/blob/9ac372568b0a63c0abb8953431fa6a392de040a3/src/librustc_mir/interpret/eval_context.rs#L553) on every loop iteration instead of once per aggregate rvalue?
* Actually enable the loop detector in [step](https://github.com/rust-lang/rust/blob/9ac372568b0a63c0abb8953431fa6a392de040a3/src/librustc_mir/interpret/step.rs#L11)
* Figure out how on earth I'm going to test for false positives :)

oli-obk commented:
The ADT value increase is fine as it is. You will have a gap in your step IDs, but that should be fine

ecstatic-morse commented:
So I've implemented, enabled and added a (successful) test for a version of the loop detector without any of the performance improvements mentioned above.

I can't tell if the `M: Machine` parameter to `EvalContext` really needs to be a part of the state: the only two implementers are [here](https://github.com/rust-lang/rust/blob/01cc982e936120acb0424e41de14e42ba2d88c6f/src/librustc_mir/interpret/const_eval.rs#L189) and [here](https://github.com/solson/miri/blob/e1734470e780e05a3366a2f74cfa25ea88a518a5/src/lib.rs#L273). The first is an empty struct and it's not clear to me why the second one is transient.
Since running the loop detector requires `M: Eq + Clone + Hash`, this PR will break `miri` until I figure out what the proper equality semantics are for it's `Machine` implementation.

Also, since the loop counter is sometimes incremented by more than one, it's possible for the loop detector to fail to when `DETECTOR_SNAPSHOT_PERIOD` equals the increment value during the course of an infinite loop. Since this is just a false negative, and pretty hard to trigger, this isn't a big deal. It will complicate some optimizations of the loop detector though. I don't quite understand why an interpreter step doesn't always equal evaluating a terminator.

Finally, as @oli-obk mentioned, it's really hard to test for false positives. It's possible that I'm not taking some part of the execution state into account, and two snapshots which compare as equal will not actually cause an infinite loop. I guess the way you usually deal with is a crater run? Not sure if that would be appropriate here...

oli-obk commented on src/test/ui/const-eval/infinite_loop.stderr:
```diff
@@ -0,0 +1,41 @@
+LL | |         let mut n = 113383; // #20 in A006884
+LL | |         while n != 0 { //~ ERROR constant contains unimplemented expression type
+LL | |             n = if n % 2 == 0 { n/2 } else { 3*n + 1 };
+   | |                    ---------- program will never terminate
```
I think it would be prudent to add more info here. Like that the highlighted expression is simply the first expression found where the constant evaluator could with certainty say that the evaluation will not terminate.

ecstatic-morse replied:
I agree, this is a bit obtuse. What about expanding the span of the error to be the entire expression? I'll need to do some more reading into the diagnostics parts of rustc to implement that. I could also replace it with something like "Infinite loop detected here, const evaluation will never terminate." But it's hard to make clear why *this* particular statement is being singled out when the problem is the whole block.

oli-obk replied:
We can't really figure out which part is at fault. I think there's no nice solution except explain more things in the message.

oli-obk commented:
> The first is an empty struct and it's not clear to me why the second one is transient. 

Well... the environment variables can be changed. So if you're writing a loop by increasing the value inside an environment variable, then you can have a loop where all state in the main memory is the same.

> Also, since the loop counter is sometimes incremented by more than one, it's possible for the loop detector to fail to when DETECTOR_SNAPSHOT_PERIOD equals the increment value during the course of an infinite loop. Since this is just a false negative and pretty hard to trigger, it's not a big deal. Being able to increment the step counter by more than one does complicate some optimizations of the loop detector though. I don't quite understand why an interpreter step doesn't always equal evaluating a terminator.

Uhm.... because that's a refactoring mess-up (by me) ;)

Just remove that call in aggregate creation and the `usize` argument to the stepping method

ecstatic-morse commented:
@oli-obk You got it!

I saw the comment:
```rust
/// Miri does not expose env vars from the host to the emulated program
```
And thought that meant that *no* environment variables would be exposed to the emulated program, but I see now that MIRI emulates environment variables.

I'm not sure what all the cases are for MIRI, but it seems like infinite loop detection is only really desirable in CTFE. Do you think I should add a way for MIRI to enable/disable this feature?

oli-obk commented:
> I'm not sure what all the cases are for MIRI, but it seems like infinite loop detection is only really desirable in CTFE. Do you think I should add a way for MIRI to enable/disable this feature?

You mean because it might end up being slow? Miri can just set the counter to `u64::max_value()`, that should not ever reach 0 xD. We can cross that bridge when it becomes a problem.

ecstatic-morse commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -46,7 +46,7 @@ pub struct EvalContext<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
     /// The number of terminators to be evaluated before enabling the infinite
     /// loop detector.
-    pub(crate) steps_until_detector_enabled: usize,
+    pub(crate) steps_until_detector_enabled: isize,
```
Yes, this could be clearer. 

My plan is to invert this counter (e.g. `steps_since_detector_enabled`), so I can use an add, then mask off the loop detector period if steps is greater than 0. The code in `inc_step_counter` currently only works with a detector period of exactly 1 due to the saturating sub. I was waiting for some clarity on the `steps` parameter to find the best way to do this.

I'm trying to make the point at which the detector comes on a comparison with `0` since it saves a cycle or two. This is probably pretty dumb :) See previous comments regarding my affinity for premature optimization.

bjorn3 commented on src/test/ui/const-eval/infinite_loop.stderr:
```diff
@@ -0,0 +1,41 @@
+LL | |         let mut n = 113383; // #20 in A006884
+LL | |         while n != 0 { //~ ERROR constant contains unimplemented expression type
+LL | |             n = if n % 2 == 0 { n/2 } else { 3*n + 1 };
+   | |                    ---------- duplicate interpreter state observed while executing this expression, const evaluation will never terminate
```
I think this label is too long.

Maybe s/while executing this expression/here

bjorn3 commented on src/test/ui/const-eval/infinite_loop.rs:
```diff
@@ -0,0 +1,25 @@
+    let _ = [(); {
+        //~^ WARNING Constant evaluating a complex constant, this might take some time
+        //~| ERROR could not evaluate repeat length
+        let mut n = 113383; // #20 in A006884
```
What is `A006884`?

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/398424563) ([raw log](https://api.travis-ci.org/v3/job/398424563/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:47:00] 
[00:47:00] running 1535 tests
[00:47:06] ................................................................................................i...
[00:47:13] .............................................................i......................................
[00:47:18] ...................................F................................................................
[00:47:25] ....................................................................................................
[00:47:28] ....................................................................................................
[00:47:34] ....................................................................................................
[00:47:38] ....................................................................................................
---
[00:48:21] ---- [ui] ui/const-eval/infinite_loop.rs stdout ----
[00:48:21] diff of stderr:
[00:48:21] 
[00:48:21] 13    |  __________________^
[00:48:21] 14 LL | |         //~^ WARNING Constant evaluating a complex constant, this might take some time
[00:48:21] 15 LL | |         //~| ERROR could not evaluate repeat length
[00:48:21] - LL | |         let mut n = 113383; // #20 in A006884
[00:48:21] + LL | |         let mut n = 113383; // #20 in https://oeis.org/A006884
[00:48:21] 17 ...  |
[00:48:21] 18 LL | |         n
[00:48:21] 19 LL | |     }];
[00:48:21] 26    |  __________________^
[00:48:21] 26    |  __________________^
[00:48:21] 27 LL | |         //~^ WARNING Constant evaluating a complex constant, this might take some time
[00:48:21] 28 LL | |         //~| ERROR could not evaluate repeat length
[00:48:21] - LL | |         let mut n = 113383; // #20 in A006884
[00:48:21] + LL | |         let mut n = 113383; // #20 in https://oeis.org/A006884
[00:48:21] 30 LL | |         while n != 0 { //~ ERROR constant contains unimplemented expression type
[00:48:21] 31 LL | |             n = if n % 2 == 0 { n/2 } else { 3*n + 1 };
[00:48:21] 32    | |                    ---------- n's value must be known at compile-time. Erroneous code\nexample:\n\n```compile_fail\nenum Test {\n    V1\n}\n\nimpl Test {\n    fn test(&self) -> i32 {\n        12\n    }\n}\n\nfn main() {\n    const FOO: Test = Test::V1;\n\n    const A: i32 = FOO.test(); // You can't call Test::func() here!\n}\n```\n\nRemember: you can't use a function call inside a const's initialization\nexpression! However, you can totally use it anywhere else:\n\n```\nenum Test {\n    V1\n}\n\nimpl Test {\n    fn func(&self) -> i32 {\n        12\n    }\n}\n\nfn main() {\n    const FOO: Test = Test::V1;\n\n    FOO.func(); // here is good\n    let x = FOO.func(); // or even here!\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/ui/const-eval/infinite_loop.rs","byte_start":878,"byte_end":1016,"line_start":20,"line_end":22,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        while n != 0 { //~ ERROR constant contains unimplemented expression type","highlight_start":9,"highlight_end":81},{"text":"            n = if n % 2 == 0 { n/2 } else { 3*n + 1 };","highlight_start":1,"highlight_end":56},{"text":"        }","highlight_start":1,"highlight_end":10}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error[E0019]: constant contains unimplemented expression type\n  --> /checkout/src/test/ui/const-eval/infinite_loop.rs:20:9\n   |\nLL | /         while n != 0 { //~ ERROR constant contains unimplemented expression type\nLL | |             n = if n % 2 == 0 { n/2 } else { 3*n + 1 };\nLL | |         }\n   | |_________^\n\n"}
[00:48:21] {"message":"Constant evaluating a complex constant, this might take some time","code":null,"level":"warning","spans":[{"file_name":"/checkout/src/test/ui/const-eval/infinite_loop.rs","byte_start":666,"byte_end":1032,"line_start":16,"line_end":24,"column_start":18,"column_end":6,"is_primary":true,"text":[{"text":"    let _ = [(); {","highlight_start":18,"highlight_end":19},{"text":"        //~^ WARNING Constant evaluating a complex constant, this might take some time","highlight_start":1,"highlight_end":87},{"text":"        //~| ERROR could not evaluate repeat length","highlight_start":1,"highlight_end":52},{"text":"        let mut n = 113383; // #20 in https://oeis.org/A006884","highlight_start":1,"highlight_end":63},{"text":"        while n != 0 { //~ ERROR constant contains unimplemented expression type","highlight_start":1,"highlight_end":81},{"text":"            n = if n % 2 == 0 { n/2 } else { 3*n + 1 };","highlight_start":1,"highlight_end":56},{"text":"        }","highlight_start":1,"highlight_end":10},{"text":"        n","highlight_start":1,"highlight_end":10},{"text":"    }];","highlight_start":1,"highlight_end":6}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"warning: Constant evaluating a complex constant, this might take some time\n  --> /checkout/src/test/ui/const-eval/infinite_loop.rs:16:18\n   |\nLL |       let _ = [(); {\n   |  __________________^\nLL | |         //~^ WARNING Constant evaluating a complex constant, this might take some time\nLL | |         //~| ERROR could not evaluate repeat length\nLL | |         let mut n = 113383; // #20 in https://oeis.org/A006884\n...  |\nLL | |         n\nLL | |     }];\n   | |_____^\n\n"}
[00:48:21] {"message":"could not evaluate repeat length","code":{"code":"E0080","explanation":"\nThis error indicates that the compiler was unable to sensibly evaluate an\nconstant expression that had to be evaluated. Attempting to divide by 0\nor causing integer overflow are two ways to induce this error. For example:\n\n```compile_fail,E0080\nenum Enum {\n    X = (1 << 500),\n    Y = (1 / 0)\n}\n```\n\nEnsure that the expressions given can be evaluated as the desired integer type.\nSee the FFI section of the Reference for more information about using a custom\ninteger type:\n\nhttps://doc.rust-lang.org/reference.html#ffi-attributes\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/ui/const-eval/infinite_loop.rs","byte_start":970,"byte_end":980,"line_start":21,"line_end":21,"column_start":20,"column_end":30,"is_primary":false,"text":[{"text":"            n = if n % 2 == 0 { n/2 } else { 3*n + 1 };","highlight_start":20,"highlight_end":30}],"label":"duplicate interpreter state observed here, const evaluation will never terminate","suggested_replacement":null,"suggestion_applicability":null,"expansion":null},{"file_name":"/checkout/src/test/ui/const-eval/infinite_loop.rs","byte_start":666,"byte_end":1032,"line_start":16,"line_end":24,"column_start":18,"column_end":6,"is_primary":true,"text":[{"text":"    let _ = [(); {","highlight_start":18,"highlight_end":19},{"text":"        //~^ WARNING Constant evaluating a complex constant, this might take some time","highlight_start":1,"highlight_end":87},{"text":"        //~| ERROR could not evaluoccurred: E0019, E0080.","code":null,"level":"","spans":[],"children":[],"rendered":"Some errors occurred: E0019, E0080.\n"}
[00:48:21] {"message":"For more information about an error, try `rustc --explain E0019`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about an error, try `rustc --explain E0019`.\n"}
[00:48:21] ------------------------------------------
[00:48:21] 
[00:48:21] thread '[ui] ui/const-eval/infinite_loop.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3140:9
[00:48:21] note: Run with `RUST_BACKTRACE=1` for a backtrace.
---
[00:48:21] test result: FAILED. 1529 passed; 1 failed; 5 ignored; 0 measured; 0 filtered out
[00:48:21] 
[00:48:21] 
[00:48:21] 
[00:48:21] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "3.9.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[00:48:21] 
[00:48:21] 
[00:48:21] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[00:48:21] Build completed unsuccessfully in 0:02:14
[00:48:21] Build completed unsuccessfully in 0:02:14
[00:48:21] Makefile:58: recipe for target 'check' failed
[00:48:21] make: *** [check] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:01741b04
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

ecstatic-morse commented:
Okay, I implemented the required traits in [`miri`](https://github.com/solson/miri/pull/387), made the changes requested by @bjorn3, and responded to @oli-obk's concern regarding using `AllocId`s for equality. There's definitely room for optimization as well as some smaller features (a hard cap on the number of iterations during const eval?), but I'm satisfied with the majority of this PR, so I'm removing the WIP tag.

Lemme know what you think!

bjorn3 commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -500,7 +624,9 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
-    ) -> EvalResult<'tcx> {
+    ) -> EvalResult<'tcx>
+        where M: Clone + Eq + Hash,
+    {
```
Maybe add this to `Machine` itself. Eg:

```
trait Machine: Clone + Eq + Hash {
[...]
```

ecstatic-morse replied:
That seems fine, `Machine` isn't really used for anything but this.

bors commented:
:umbrella: The latest upstream changes (presumably #51428) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented:
> As long as MIRI only executes pure functions, we can detect if a program will never terminate by periodically taking a "snapshot" of this transient state and comparing it to previous ones. If any two states are exactly equal, the machine must be in an infinite loop.

However, it is possible to diverge without ever repeating a state -- if the program just has larger and larger states. So you are not reliably detecting all cases of programs never terminating, you just detect some cases. (Which you may be aware of, but the way you wrote this made it sound like you are claiming more than this can possibly do. :D )

oli-obk commented:
We do detect nontermination. It might take us forever or the system ooms before we do, but in all other cases, at some point we've seen all possible States. Oom also happens faster due to the hashset building up.

On a more serious note: as long as we detect simple nontermination it's fine. The complex cases will still produce a warning, so users know something's off. Rustc already isn't guaranteed to terminate in it's typesystem and macro system, so...

ecstatic-morse commented:
I changed the PR description to "we can detect if a program is in a state where it will never terminate" to avoid accusations that I've claimed to solve the halting problem.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/400154080) ([raw log](https://api.travis-ci.org/v3/job/400154080/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:07:41]    Compiling proc_macro v0.0.0 (file:///checkout/src/libproc_macro)
[00:08:01]    Compiling syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)
[00:13:56]    Compiling rustc_typeck v0.0.0 (file:///checkout/src/librustc_typeck)
[00:13:56]    Compiling rustc_mir v0.0.0 (file:///checkout/src/librustc_mir)
[00:13:57] error[E0432]: unresolved import `rustc_data_structures::fx::FxHashxHasher`
[00:13:57]   --> librustc_mir/interpret/memory.rs:13:55
[00:13:57]    |
[00:13:57] 13 | use rustc_data_structures::fx::{FxHashSet, FxHashMap, FxHashxHasher};
[00:13:57]    |                                                       ^^^^^^^^^^^^^ no `FxHashxHasher` in `fx`
[00:13:57] 
[00:13:58] error[E0433]: failed to resolve. Use of undeclared type or module `FxHasher`
[00:13:58]    --> librustc_mir/interpret/memory.rs:104:29
[00:13:58]     |
[00:13:58] 104 |                 let mut h = FxHasher::default();
[00:13:58]     |                             ^^^^^^^^ Use of undeclared type or module `FxHasher`
[00:13:58] error: unused import: `ptr`
[00:13:58] error: unused import: `ptr`
[00:13:58]  --> librustc_mir/interpret/eval_context.rs:3:16
[00:13:58]   |
[00:13:58] 3 | use std::{mem, ptr};
[00:13:58]   |
[00:13:58]   = note: `-D unused-imports` implied by `-D warnings`
[00:13:58] 
[00:13:58] 
[00:13:58] error: unused import: `FxHashxHasher`
[00:13:58]   --> librustc_mir/interpret/memory.rs:13:55
[00:13:58]    |
[00:13:58] 13 | use rustc_data_structures::fx::{FxHashSet, FxHashMap, FxHashxHasher};
[00:13:58] 
[00:13:58] 
[00:14:08] error[E0026]: struct `interpret::memory::Memory` does not have a field named `uninitialized_statics`
[00:14:08]   --> librustc_mir/interpret/memory.rs:66:13
[00:14:08] 66 |             uninitialized_statics,
[00:14:08] 66 |             uninitialized_statics,
[00:14:08]    |             ^^^^^^^^^^^^^^^^^^^^^ struct `interpret::memory::Memory` does not have this field
[00:14:08] 
[00:14:08] error[E0609]: no field `uninitialized_statics` on type `&interpret::memory::Memory<'a, 'mir, 'tcx, M>`
[00:14:08]   --> librustc_mir/interpret/memory.rs:74:48
[00:14:08]    |
[00:14:08] 74 |             && *uninitialized_statics == other.uninitialized_statics
[00:14:08] 
[00:14:08] 
[00:14:08] error[E0026]: struct `interpret::memory::Memory` does not have a field named `uninitialized_statics`
[00:14:08]   --> librustc_mir/interpret/memory.rs:88:13
[00:14:08] 88 |             uninitialized_statics: _,
[00:14:08] 88 |             uninitialized_statics: _,
[00:14:08]    |             ^^^^^^^^^^^^^^^^^^^^^^^^ struct `interpret::memory::Memory` does not have this field
[00:14:10] error: aborting due to 7 previous errors
[00:14:10] 
[00:14:10] Some errors occurred: E0026, E0432, E0433, E0609.
[00:14:10] For more information about an error, try `rustc --explain E0026`.
[00:14:10] For more information about an error, try `rustc --explain E0026`.
[00:14:10] error: Could not compile `rustc_mir`.
[00:14:10] 
[00:14:10] Caused by:
[00:14:10]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc_mir librustc_mir/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=8fc603afb8ea9b13 -C extra-filename=-8fc603afb8ea9b13 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libarena-d45628fe21047b42.so --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-d8b3f1986e621085.rlib --extern byteorder=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbyteorder-09e9dbd1ef48ffa5.rlib --extern either=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libeither-b51deb005c05dd3b.rlib --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libgraphviz-6020508f01da724d.so --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-023d781fbd65d983.rlib --extern log_settings=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog_settings-9c861d36e123bec8.rlib --extern polonius_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libpolonius_engine-2e546cbdf217aece.rlib --extern rustc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc-1dddb0fa9d8a512f.so --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-be9737b074a8dae0.rlib --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-25582ce13d3618ea.so --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-e036f8f5b9204e52.so --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-066098b54e835a1f.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-05774896829b5d87.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-05774896829b5d87.rlib --extern syntax=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax-a1b02e0d020520ac.so --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-f36f6cb494d373be.so -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/backtrace-sys-a59a4023b81a89b2/out -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/miniz-sys-ea5907e223517cf2/out` (exit code: 101)
travis_time:end:05800cb3:start=1530736414669097161,finish=1530737361683726248,duration=947014629087

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:10f5fe94
---
travis_time:end:03cafb18:start=1530737362026617252,finish=1530737362033513137,duration=6895885
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:1b2a13c8
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:07ae6e18
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

oli-obk commented on src/test/ui/const-eval/infinite_loop_hard.rs:
```diff
@@ -0,0 +1,27 @@
+    &x
+}
+
+fn main() {
```
Uh. We don't detect this loop. So our test suite will now not terminate? XD

Please remove this test

ecstatic-morse replied:
How did that get there? An errant `stash apply` I think.

oli-obk commented on src/test/ui/const-eval/infinite_loop.rs:
```diff
@@ -0,0 +1,25 @@
+// except according to those terms.
+
+#![feature(const_let)]
+
```
Just out of curiosity, how long does this test run before finishing with an error?

ecstatic-morse replied:
I believe a second or two, I'll rerun it to make sure.

ecstatic-morse replied:
> finished in 1.492

oli-obk replied:
For now I think it's fine. We should probably run a perf run first though

oli-obk replied:
If it is a problem, we can try some tricks like exponential backoff paired with clearing the HashSet

oli-obk commented:
r=me with halting problem solved or test removed

ecstatic-morse commented on src/libcore/ptr.rs:
```diff
@@ -2813,6 +2814,7 @@ impl<'a, T: ?Sized> From<NonNull<T>> for Unique<T> {
 /// such as Box, Rc, Arc, Vec, and LinkedList. This is the case because they
 /// provide a public API that follows the normal shared XOR mutable rules of Rust.
 #[stable(feature = "nonnull", since = "1.25.0")]
+#[repr(transparent)]
```
Well this is embarrassing.

ecstatic-morse replied:
Should be fixed, thanks for double checking. I rebased to fix a conflict and must have somehow based it on the wrong commit.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit cf5eaa75bb171f00d6baa475333b741b86f93f72 has been approved by `oli-obk`

<!-- @bors r=oli-obk cf5eaa75bb171f00d6baa475333b741b86f93f72 -->

oli-obk commented:
@bors r-

oli-obk commented:
@bors try

bors commented:
:hourglass: Trying commit cf5eaa75bb171f00d6baa475333b741b86f93f72 with merge cf47daa968f12c4f95111ce03aaf02b734cc78d7...

oli-obk commented:
@Mark-Simulacrum can you start a perf run on this?

bors commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang/rust/builds/400187326?utm_source=github_status&utm_medium=notification)
State: approved= try=True

RalfJung commented:
> We do detect nontermination. It might take us forever or the system ooms before we do, but in all other cases, at some point we've seen all possible States. Oom also happens faster due to the hashset building up.

Oh I see -- you're kind of "cheating" here because you rely on finite host machine memory to implicitly limit the state space of the virtual machine. Fair enough ;)

Mark-Simulacrum commented:
@rust-timer build cf47daa968f12c4f95111ce03aaf02b734cc78d7

Also added you to permissions.

rust-timer commented:
Success: Queued cf47daa968f12c4f95111ce03aaf02b734cc78d7 with parent 4af9132a022ebe4111395139d1a4e45e285d299a, [comparison URL](https://perf.rust-lang.org/compare.html?start=4af9132a022ebe4111395139d1a4e45e285d299a&end=cf47daa968f12c4f95111ce03aaf02b734cc78d7).

kennytm commented:
Perf is now ready. The timing changes are... mixed 😕  (inflate-opt @ +2.4% and syn-opt @ -1.0% are the two extremes).

Mark-Simulacrum commented:
Probably just noise, those seem to be fairly normal measurements for most noop builds.

ecstatic-morse commented:
I recorded the benchmark locally with `perf`, and `interpret::step` and its children in the call graph were only 0.01% of total cycles. I doubt that `inflate` is actually triggering the detector; I think it would take up more time. I'm not sure how common spurious data points are, especially since you're counting instructions, not CPU cycles, but the minimum run for `inflate` is -0.1%, so there must be some variance there.

oli-obk commented:
@bors r+

We do emit a warning if things start going slow, so at least ppl will know what's going on.

bors commented:
:pushpin: Commit cf5eaa75bb171f00d6baa475333b741b86f93f72 has been approved by `oli-obk`

<!-- @bors r=oli-obk cf5eaa75bb171f00d6baa475333b741b86f93f72 -->

bors commented:
:hourglass: Testing commit cf5eaa75bb171f00d6baa475333b741b86f93f72 with merge d6364f49d822ba1c5076ac649327c048bc34a96e...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/402528368?utm_source=github_status&utm_medium=notification)

rust-highfive commented:
The job `x86_64-gnu-nopt` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/402528409) ([raw log](https://api.travis-ci.org/v3/job/402528409/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
#######                                                                   11.1%
######################################################################## 100.0%
[00:01:03] extracting /checkout/obj/build/cache/2018-06-30/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:01:04]     Updating registry `https://github.com/rust-lang/crates.io-index`
No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-nopt` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/402528409) ([raw log](https://api.travis-ci.org/v3/job/402528409/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
#######                                                                   11.1%
######################################################################## 100.0%
[00:01:03] extracting /checkout/obj/build/cache/2018-06-30/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:01:04]     Updating registry `https://github.com/rust-lang/crates.io-index`
No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

oli-obk commented:
@bors retry

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.8106), [status-travis](https://travis-ci.org/rust-lang/rust/builds/402746043?utm_source=github_status&utm_medium=notification)
Approved by: oli-obk
Pushing d573fe17786b9c2f5a766498d411d54eee5fa19f to master...
